components/proto/src/funder/messages.rs

```diff
@@ -308,68 +310,77 @@ impl FriendsRoute {
+        hash::sha_512_256(&self.canonical_serialize())
     }
 
-    /// Check if the route is valid.
-    /// A valid route must have at least 2 unique nodes, and is in one of the following forms:
-    /// A -- B -- C -- D -- E -- F -- A   (Single cycle, first == last)
-    /// A -- B -- C -- D -- E -- F        (A route with no repetitions)
-    pub fn is_valid(&self) -> bool {
-        if self.public_keys.len() < 2 {
-            return false;
-        }
-        if self.public_keys.len() > MAX_ROUTE_LEN {
-            return false;
-        }
+    /// Get the public key of a node according to its index.
+    pub fn index_to_pk(&self, index: usize) -> Option<&PublicKey> {
+        self.public_keys.get(index)
+    }
+}
 
-        let last_key = &self.public_keys[self.public_keys.len() - 1];
-        if last_key == &self.public_keys[0] {
-            // We have a first == last cycle.
-            if self.public_keys.len() > 2 {
-                // We have a cycle that is long enough (no A -- A).
-                // We just check if it's a single cycle.
-                Self::is_unique(&self.public_keys[1..])
-            } else {
-                // A -- A
-                false
-            }
-        } else {
-            // No first == last cycle.
-            // But we have to check if there is any other cycle.
-            Self::is_unique(&self.public_keys)
-        }
+use std::ops::Deref;
+/// This `Deref` lets us use `is_route_valid` over `FriendsRoute`
+impl Deref for FriendsRoute {
+    type Target = [PublicKey];
+    fn deref(&self) -> &Self::Target {
+        self.public_keys.as_ref()
     }
+}
 
-    /// Checks if the remaining part of the route is valid.
-    /// Compared to regular version, this one does not check for minimal unique
-    /// nodes amount. It returns `true` if the part is empty.
-    /// It does not accept routes parts with a cycle, though.
-    pub fn is_valid_part(&self) -> bool {
-        if self.public_keys.len() > MAX_ROUTE_LEN - 1 {
+/// Check if no element repeats twice in the slice
+fn no_duplicates<T: Hash + Eq>(array: &[T]) -> bool {
+    let mut seen = HashSet::new();
+    for item in array {
+        if !seen.insert(item) {
             return false;
         }
+    }
+    true
+}
 
-        Self::is_unique(&self.public_keys)
+/// Check if the route (e.g. `FriendsRoute`) is valid.
+/// A valid route must have at least 2 unique nodes, and is in one of the following forms:
+/// A -- B -- C -- D -- E -- F -- A   (Single cycle, first == last)
+/// A -- B -- C -- D -- E -- F        (A route with no repetitions)
+pub fn is_route_valid<T: Hash + Eq>(route: &[T]) -> bool {
+    if route.len() < 2 {
+        return false;
+    }
+    if route.len() > MAX_ROUTE_LEN {
+        return false;
     }
 
-    /// Produce a cryptographic hash over the contents of the route.
-    pub fn hash(&self) -> HashResult {
-        hash::sha_512_256(&self.canonical_serialize())
+    let last_key = &route[route.len() - 1];
```

Comment:
You might be able to use `.last()` here.
My reasoning is that I am super afraid of subtraction and addition operations (On release they can overflow). However, using `last()` requires you to `unwrap()` to get the last item. I rather have a panic than an overflow.

Reply:
In the current code it cannot fail as we have `len() >= 2` guarantee. But it's not a bad idea, I'll add it in my next commit.